Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1

```diff
@@ -380,6 +380,157 @@ function Wait-ServiceFabricApplicationTypeStatusChange
+
+    $unregisterAction = { Unregister-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -Force -TimeoutSec $TimeoutSec }
+
+    $exceptionRetryEvaluator = {
+        param($ex)
+        $appType = Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+        # If app type already unprovisioned, don't retry
+        if (!$appType)
+        {
+            return $false
+        }
+
+        # if unprovisioning started, wait for it to complete
+        if (($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -or ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning))
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeUnprovisioningStarted)
+            $appType = Wait-ServiceFabricApplicationTypeStatusChange -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+        }
+
+        # if app type got unprovisioned, don't retry
+        if (!$appType)
+        {
+            return $false
+        }
+
+        # if app type is still provisioned, retry unregister
+        if ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Available)
+        {
+            return $true
+        }
+
+        # if unprovisioning failed, throw and don't retry
+        throw (Get-VstsLocString -Key SFSDK_UnregisterAppTypeFailedWithStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
+    }
+
+    try
+    {
+        Invoke-ActionWithDefaultRetries -Action $unregisterAction `
+            -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingUnregisterApplicationType) `
+            -ExceptionRetryEvaluator $exceptionRetryEvaluator
+    }
+    catch
+    {
+        # print cluster health status if unregistering failed
+        Trace-ServiceFabricClusterHealth
+        throw
+    }
+}
+
+function Remove-ServiceFabricApplicationAction
+{
+    Param(
+        [string]
+        $ApplicationName
+    )
+
+    $global:operationId = $SF_Operations.RemoveApplication
+
+    $removeAction = { Remove-ServiceFabricApplication -ApplicationName $ApplicationName -Force }
+    $exceptionRetryEvaluator = {
+        param($ex)
+
+        # If app already removed, don't retry
+        if ($ex.GetType().FullName -eq "System.Fabric.FabricElementNotFoundException")
+        {
+            return $false
+        }
+
+        return $true
+    }
+
+    try
+    {
+        Invoke-ActionWithDefaultRetries -Action $removeAction `
+            -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingRemoveApplication) `
+            -ExceptionRetryEvaluator $exceptionRetryEvaluator `
+            -RetryableExceptions @("System.Fabric.FabricTransientException", "System.Fabric.FabricElementNotFoundException", "System.TimeoutException")
+    }
+    catch [System.TimeoutException]
```

Comment:
You're passing System.TimeoutException as a retryable exception in function Invoke-ActionWithDefaultRetries. So this catch will activated only after 3 retries. Why are we doing 3 retries on the first place for this exception when we know that we need to handle this exception in specific logic (logic mentioned in catch).
Suggestion: 
Add following code in $exceptionRetryEvaluator:
if ($ex.GetType().FullName -eq "System.TimeoutException")
{
       throw $ex;
}

Reply:
we are re-trying remove action in case of timeout - just like all other commands. may be force will not be required if we retry timeout. Are you sure that one timeout means force remove?

Reply:
I guess thats the current behavior. Why do you want to change the behavior? This will unnecessary add 30 seconds in the process which is HUGE!!